File: frontend/src/api/workspaces.ts
import {
  BookConfig,
  BookCreateRequest,
  BookSnapshot,
  BookSummary,
  BookUpdateRequest,
  CollectionCreateRequest,
  CollectionSummary,
  CollectionUpdateRequest,
  WorkspaceAvailabilityResponse,
  WorkspaceComparison,
  WorkspaceComparisonRequest,
  WorkspaceDetail,
  WorkspaceSummary,
  WorkspaceUpdateRequest,
  WorkspaceView,
} from "./types";
import { request } from "./request";

export const workspaceApi = {
  listWorkspaces: () => request<WorkspaceSummary[]>("/workspaces"),
  createWorkspace: (payload: { start_date: string }) =>
    request<WorkspaceDetail>("/workspaces", {
      method: "POST",
      body: JSON.stringify(payload),
    }),
  updateWorkspace: (workspaceId: string, payload: WorkspaceUpdateRequest) =>
    request<WorkspaceDetail>(`/workspaces/${workspaceId}`, {
      method: "PATCH",
      body: JSON.stringify(payload),
    }),
  getWorkspace: (workspaceId: string) => request<WorkspaceDetail>(`/workspaces/${workspaceId}`),
  getWorkspaceView: (workspaceId: string, signal?: AbortSignal) =>
    request<WorkspaceView>(`/workspaces/${workspaceId}/view`, { signal }),
  deleteWorkspace: (workspaceId: string) =>
    request<void>(`/workspaces/${workspaceId}`, { method: "DELETE" }),
  createCollection: (workspaceId: string, payload: CollectionCreateRequest) =>
    request<CollectionSummary>(`/workspaces/${workspaceId}/collections`, {
      method: "POST",
      body: JSON.stringify(payload),
    }),
  updateCollection: (collectionId: string, payload: CollectionUpdateRequest) =>
    request<CollectionSummary>(`/collections/${collectionId}`, {
      method: "PATCH",
      body: JSON.stringify(payload),
    }),
  deleteCollection: (collectionId: string) =>
    request<void>(`/collections/${collectionId}`, { method: "DELETE" }),
  listBooks: (workspaceId: string) => request<BookSummary[]>(`/workspaces/${workspaceId}/books`),
  createBook: (collectionId: string, payload: BookCreateRequest) =>
    request<BookSummary>(`/collections/${collectionId}/books`, {
      method: "POST",
      body: JSON.stringify(payload),
    }),
  getBookConfig: (bookId: string, signal?: AbortSignal) => request<BookConfig>(`/books/${bookId}/config`, { signal }),
  updateBook: (bookId: string, payload: BookUpdateRequest) =>
    request<BookSummary>(`/books/${bookId}`, {
      method: "PATCH",
      body: JSON.stringify(payload),
    }),
  deleteBook: (bookId: string) => request<void>(`/books/${bookId}`, { method: "DELETE" }),
  getWorkspaceAvailability: (workspaceId: string, tickers: string[], signal?: AbortSignal) =>
    request<WorkspaceAvailabilityResponse>(`/workspaces/${workspaceId}/availability`, {
      method: "POST",
      signal,
      body: JSON.stringify({ tickers }),
    }),
  getWorkspaceComparison: (workspaceId: string, payload: WorkspaceComparisonRequest, signal?: AbortSignal) =>
    request<WorkspaceComparison>(`/workspaces/${workspaceId}/comparison`, {
      method: "POST",
      signal,
      body: JSON.stringify(payload),
    }),
  getBookSnapshot: (bookId: string, asOf: string, benchmarkTicker?: string | null, signal?: AbortSignal) => {
    const params = new URLSearchParams({ as_of: asOf });
    if (benchmarkTicker) {
      params.set("benchmark_ticker", benchmarkTicker);
    }
    return request<BookSnapshot>(`/books/${bookId}/snapshot?${params.toString()}`, { signal });
  },
};

File: backend/routers/workspaces.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Response, status
from sqlalchemy.orm import Session

from backend.database import get_db
from backend.models.schemas import (
    BookSummary,
    WorkspaceAvailabilityRequest,
    WorkspaceAvailabilityResponse,
    WorkspaceComparison,
    WorkspaceComparisonRequest,
    WorkspaceCreate,
    WorkspaceDetail,
    WorkspaceSummary,
    WorkspaceUpdate,
    WorkspaceView,
)
from backend.services.workspace_service import WorkspaceService

router = APIRouter(tags=["workspaces"])


@router.get("/workspaces", response_model=list[WorkspaceSummary])
def list_workspaces(db: Session = Depends(get_db)) -> list[WorkspaceSummary]:
    return WorkspaceService(db).list_workspaces()


@router.post("/workspaces", response_model=WorkspaceDetail, status_code=status.HTTP_201_CREATED)
def create_workspace(payload: WorkspaceCreate, db: Session = Depends(get_db)) -> WorkspaceDetail:
    return WorkspaceService(db).create_workspace(payload)


@router.get("/workspaces/{workspace_id}", response_model=WorkspaceDetail)
def get_workspace(workspace_id: str, db: Session = Depends(get_db)) -> WorkspaceDetail:
    return WorkspaceService(db).build_workspace_detail(workspace_id)


@router.get("/workspaces/{workspace_id}/view", response_model=WorkspaceView)
def get_workspace_view(workspace_id: str, db: Session = Depends(get_db)) -> WorkspaceView:
    return WorkspaceService(db).build_workspace_view(workspace_id)


@router.patch("/workspaces/{workspace_id}", response_model=WorkspaceDetail)
def update_workspace(payload: WorkspaceUpdate, workspace_id: str, db: Session = Depends(get_db)) -> WorkspaceDetail:
    return WorkspaceService(db).update_workspace(workspace_id, payload)


@router.delete("/workspaces/{workspace_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_workspace(workspace_id: str, db: Session = Depends(get_db)) -> Response:
    WorkspaceService(db).delete_workspace(workspace_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/workspaces/{workspace_id}/books", response_model=list[BookSummary])
def list_books(workspace_id: str, db: Session = Depends(get_db)) -> list[BookSummary]:
    return WorkspaceService(db).list_books(workspace_id)


@router.post("/workspaces/{workspace_id}/comparison", response_model=WorkspaceComparison)
def get_workspace_comparison(
    workspace_id: str,
    payload: WorkspaceComparisonRequest,
    db: Session = Depends(get_db),
) -> WorkspaceComparison:
    return WorkspaceService(db).build_comparison(workspace_id, payload)


@router.post("/workspaces/{workspace_id}/availability", response_model=WorkspaceAvailabilityResponse)
def get_workspace_availability(
    workspace_id: str,
    payload: WorkspaceAvailabilityRequest,
    db: Session = Depends(get_db),
) -> WorkspaceAvailabilityResponse:
    return WorkspaceService(db).build_workspace_availability(workspace_id, payload.tickers)
